### PR TITLE
CSHARP-3001: RetryableWrites specification improvements.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5e83d4131e2d170c181486b7

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs [hunk 1]
```diff
@@ -86,9 +86,16 @@ static RetryabilityHelper()
...
-        public static void AddRetryableWriteErrorLabelIfRequired(MongoException exception)
+        public static void AddRetryableWriteErrorLabelIfRequired(MongoException exception, SemanticVersion serverVersion)
         {
-            if (ShouldRetryableWriteExceptionLabelBeAdded(exception))
+            if (!Feature.RetryableWrites.IsSupported(serverVersion))
+            {
+                return;
+            }
+
+            if (IsNetworkException(exception) ||
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs]: according to spec, the driver should add an error label for network errors regardless server version (if >=3.6)

File: tests/MongoDB.Driver.Tests/RetryableWritesTests.cs [hunk 2]
```diff
@@ -43,6 +45,34 @@ public void Insert_with_RetryWrites_true_should_work_whether_retryable_writes_ar
             }
         }
 
+        [SkippableFact]
+        public void Retryable_write_errorlabel_should_not_be_added_with_retryWrites_false()
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/RetryableWritesTests.cs:49]: Ensure that error label won't be added for operations without retryWrites

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs [hunk 3]
```diff
@@ -179,11 +179,9 @@ private void AddErrorLabelIfRequired(Exception exception, SemanticVersion server
                     mongoException.AddErrorLabel("TransientTransactionError");
                 }
 
-                if ((exception is MongoConnectionException) || // network error
-                    (Feature.RetryableWrites.IsSupported(serverVersion) &&
-                    !Feature.ServerReturnsRetryableWriteErrorLabel.IsSupported(serverVersion)))
+                if (_command.Contains("txnNumber")) // operation is retryable
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs]: See for details: https://github.com/mongodb/specifications/blob/49992a6d1c14b36421e778e78070a427887674fb/source/retryable-writes/retryable-writes.rst#why-does-the-driver-only-add-the-retryablewriteerror-label-to-errors-that-occur-on-a-mongoclient-with-retrywrites-set-to-true

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs]: Put all these new conditions inside `ShouldRetryableWriteExceptionLabelBeAdded`.

Change old line 91 (new lines 91-98) to:

    if (ShouldRetryableWriteExceptionLabelBeAdded(exception, serverVersion))

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs [hunk 4]
```diff
@@ -45,18 +45,24 @@ public void AddRetryableWriteErrorLabelIfRequired_should_add_RetryableWriteError
...
         {
             var exception = (MongoException)CoreExceptionHelper.CreateException(typeof(MongoConnectionException));
 
-            RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(exception);
+            var serverReturnsRetryableWriteErrorLabelFeature = Feature.ServerReturnsRetryableWriteErrorLabel;
+            RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(
+                exception, 
+                isNewBehavior 
+                    ? serverReturnsRetryableWriteErrorLabelFeature.FirstSupportedVersion
+                    : serverReturnsRetryableWriteErrorLabelFeature.LastNotSupportedVersion);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs]: Move argument setup to the Arrange section keeping the Act section as simple as possible. Refactor lines 58-68 to:

    var exception = (MongoException)CoreExceptionHelper.CreateException(typeof(MongoConnectionException));
    var feature = Feature.ServerReturnsRetryableWriteErrorLabel;
    var serverVersion = serverReturnsRetryableWriteErrorLabel ? feature.FirstSupportedVersion : feature.LastNotSupportedVersion;

    RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(exception, serverVersion);

    exception.HasErrorLabel("RetryableWriteError").Should().BeTrue();

File: tests/MongoDB.Driver.Tests/RetryableWritesTests.cs [hunk 5]
```diff
@@ -43,6 +45,34 @@ public void Insert_with_RetryWrites_true_should_work_whether_retryable_writes_ar
...
+                    errorCode : 262
+                }
+            }"; // no error label
+
+            using (var client = GetClient(retryWrites: false))
+            using (ConfigureFailPoint(failPointWithRetryableError))
+            {
+                var database = client.GetDatabase(DriverTestConfiguration.DatabaseNamespace.DatabaseName);
+                var collection = database.GetCollection<BsonDocument>(DriverTestConfiguration.CollectionNamespace.CollectionName);
+                var document = new BsonDocument("x", 1);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/RetryableWritesTests.cs:81]: I would add a blank line between lines 68 and 69.

Line 68 is the last line of the Arrange section and line 69 is the Act section (and lines 71-72 are teh Assert section).

**Review by rstam — CHANGES_REQUESTED** [review]
Looks good. Only minor changes requested.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs]: done

File: tests/MongoDB.Driver.Tests/RetryableWritesTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/RetryableWritesTests.cs:81]: done

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs]: done

**Review by DmitryLukyanov — COMMENTED** [review]
Ready for review
EG: https://evergreen.mongodb.com/version/5e8a3b2932f417510b77907e

**Review by rstam — APPROVED** [review]
LGTM

**Comment by DmitryLukyanov** [comment]
Sorry guys, I didn't notice that some transaction tests fail on EG for old servers. I've created a driver ticket to clarify what we should do to solve them:
https://jira.mongodb.org/browse/DRIVERS-1181
Until resolving of this ticket, I will remove reviewers from this ticket :(

**Review by rstam — CHANGES_REQUESTED** [review]
Waiting for new commits

**Comment by DmitryLukyanov** [comment]
ready for review: https://evergreen.mongodb.com/version/5e924c69a4cf475229cd75be

Not sure that the solution in the latest commit is good enough, but it's the simplest. Thoughts?

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs [hunk 6]
```diff
@@ -86,17 +87,20 @@ static RetryabilityHelper()
...
                 exception.AddErrorLabel(RetryableWriteErrorLabel);
             }
         }
 
-        public static bool IsNetworkException(Exception exception)
+        public static bool IsCommandRetryable(BsonDocument command)
         {
-            return exception is MongoConnectionException mongoConnectionException && mongoConnectionException.IsNetworkException;
+            return
+                command.Contains("txnNumber") || // retryWrites=true
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: AFAIK a command is not retryable if it is in a transaction.

Do we need to check here whether the session is in a transaction or not?

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs [hunk 7]
```diff
@@ -179,11 +179,9 @@ private void AddErrorLabelIfRequired(Exception exception, SemanticVersion server
                     mongoException.AddErrorLabel("TransientTransactionError");
                 }
 
-                if ((exception is MongoConnectionException) || // network error
-                    (Feature.RetryableWrites.IsSupported(serverVersion) &&
-                    !Feature.ServerReturnsRetryableWriteErrorLabel.IsSupported(serverVersion)))
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: Can you elaborate on this completely different test?

My best guess is that before we would try to determine if the operation *should* have been retryable, but now we are looking inside the actual command to see if it *actually* is retryable?

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 6 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: p.s. not sure how you would know here if we were in a transaction or not.

**Review by rstam — CHANGES_REQUESTED** [review]
No changes requested. Just questions.

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: yes, the command is not retryable if it's within transaction, but commitTransaction and abortTransaction are retryable

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: >Do we need to check here whether the session is in a transaction or not?

I can add this validation, but it's not 100% required since we cannot have txtNumber is being filled if the operation within a transaction based on this logic: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Operations/RetryableWriteOperationExecutor.cs#L149

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: Let me know what do you think of whether this logic is required or not. If so, it will look similar to:

        public static bool IsCommandRetryable(BsonDocument command, bool isInTransaction)
        {
            return
                ((command.Contains("txnNumber") && !isInTransaction)) || // retryWrites=true
                command.Contains("commitTransaction") ||
                command.Contains("abortTransaction");
        }

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: In the previous commit: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/123/commits/7694a47c327e1107f4aa836e0f076498b81f83a9#diff-91a7ef566b7da37c92d9773b00943b16R182 we had this logic:

    if (_command.Contains("txnNumber"))
    {
      RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(mongoException, serverVersion);
    }

where the validation on txtNumber meant that we want to add the retryable write label only for operations with configured `retryWrites=true`. But we also need to add this label for commitTransaction and abortTransaction which are also retryable regardless the retryWrites option. Spec:

>Drivers MUST retry the commitTransaction and abortTransaction commands even when retryWrites has been disabled on the MongoClient

https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#interaction-with-retryable-writes

**Comment by DmitryLukyanov** [comment]
answered on Robert's questions

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs (hunk 7 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: Instead of comparing to the last commit can you compare the changes to master?

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: The whole logic from the master was moved inside `RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired`, so no real changes in this part. 
In the previous commit, in addition to the conditions that are already in master, I've added the txtNumber condition to add the write error label only for operations with configured `retryWrites=true`. In the current version, additionally to this condition, I've added two more: 
* on commitTransaction
* on abortTransaction
because these operations are retryable regardless of `retryWrites=true`.

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: as I wrote here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/123#issuecomment-612541406, It can be not the best approach, but it's the simplest

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: I was hoping you would summarize the change in one place here.

I will follow all your links and see if I can figure out what you are saying.

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: I think I summarized it :) , the whole point that now we have 3 more additional conditions for the logic which is in master now.
These 3 conditions are:
* check that retryWrites=true is configured
* the operation is commitTransaction
* the operation is abortTransaction

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: I don't think your statement is true. 

We don't have 3 ADDITIONAL conditions. What we have is 3 NEW conditions that completely REPLACE the previous conditions.

This is what I said before: My best guess is that before we would try to determine if the operation *should* have been retryable, but now we are looking inside the actual command to see if it *actually* is retryable?

Is what I said true? If not, can you summarize why you completely REPLACED the logic with new logic?

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: >We don't have 3 ADDITIONAL conditions. What we have is 3 NEW conditions that completely REPLACE the previous conditions.

no, it's not true. As I wrote here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/123#discussion_r408103022 the previous comparing logic based on server versions and error codes was moved inside `RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired`. See: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/123/files#diff-f3cf36bc2821907b3003ced21fae1593R166-R181
Now, the selected lines contain only new conditions based on the operation name and operation options.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: Oh that's tricky... you didn't move them to `RetryabilityHelper.IsCommandRetryable`

You moved them to `RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired`

No wonder I didn't realize what you had done.

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: :)

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 6 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:101]: I think it's OK to assume that if the command contains "txnNumber" that we know that is not in a transaction.

Thanks for the clarification.

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs (hunk 7 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs:184]: I'm going to resolve this comment. Thanks for the explanations, even if they didn't make sense to me at first.

**Review by rstam — APPROVED** [review]
LGTM

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs [hunk 8]
```diff
@@ -86,12 +89,46 @@ public void AddRetryableWriteErrorLabelIfRequired_should_add_RetryableWriteError
...
-            RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(exception);
+            RetryabilityHelper.AddRetryableWriteErrorLabelIfRequired(exception, Feature.ServerReturnsRetryableWriteErrorLabel.LastNotSupportedVersion);
 
             var hasRetryableWriteErrorLabel = exception.HasErrorLabel("RetryableWriteError");
             hasRetryableWriteErrorLabel.Should().Be(shouldAddErrorLabel);
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void AddRetryableWriteErrorLabelIfRequired_should_not_add_error_label_for_non_retryWrites_server([Values(false, true)] bool isNetworkError)
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs]: Style nit: Move parameters to new line? :)

**Review by vincentkam — APPROVED** [review]
Super minor style nit, otherwise LGTM! 😸(●'◡'●)

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/RetryabilityHelperTests.cs]: done
